hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/checksum/ECFileChecksumHelper.java

```diff
@@ -0,0 +1,211 @@
+    int bytesPerChecksum = checksumData.getBytesPerChecksum();
+    setBytesPerCRC(bytesPerChecksum);
+
+    ByteBuffer blockChecksumByteBuffer =
+        getBlockChecksumFromChunkChecksums(chunkInfos);
+    String blockChecksumForDebug =
+        populateBlockChecksumBuf(blockChecksumByteBuffer);
+
+    LOG.debug("Got reply from EC pipeline {} for block {}: blockChecksum={}, " +
+            "blockChecksumType={}",
+        keyLocationInfo.getPipeline(), keyLocationInfo.getBlockID(),
+        blockChecksumForDebug, checksumData.getType());
+    return true;
+  }
+
+  private String populateBlockChecksumBuf(
+      ByteBuffer blockChecksumByteBuffer) throws IOException {
+    String blockChecksumForDebug = null;
+    switch (getCombineMode()) {
+    case MD5MD5CRC:
+      final MD5Hash md5 = new MD5Hash(blockChecksumByteBuffer.array());
+      md5.write(getBlockChecksumBuf());
+      if (LOG.isDebugEnabled()) {
+        blockChecksumForDebug = md5.toString();
+      }
+      break;
+    case COMPOSITE_CRC:
+      byte[] crcBytes = blockChecksumByteBuffer.array();
+      if (LOG.isDebugEnabled()) {
+        blockChecksumForDebug = CrcUtil.toSingleCrcString(crcBytes);
+      }
+      getBlockChecksumBuf().write(crcBytes);
+      break;
+    default:
+      throw new IOException(
+          "Unknown combine mode: " + getCombineMode());
+    }
+
+    return blockChecksumForDebug;
+  }
+
+  private ByteBuffer getBlockChecksumFromChunkChecksums(
+      List<ContainerProtos.ChunkInfo> chunkInfos) throws IOException {
+
+    AbstractBlockChecksumComputer blockChecksumComputer =
+        new ECBlockChecksumComputer(chunkInfos, getKeyInfo());
+    blockChecksumComputer.compute(getCombineMode());
+
+    return blockChecksumComputer.getOutByteBuffer();
+  }
+
+  private List<ContainerProtos.ChunkInfo> getChunkInfos(OmKeyLocationInfo
+      keyLocationInfo) throws IOException {
+    // To read an EC block, we create a STANDALONE pipeline that contains the
+    // single location for the block index we want to read. The EC blocks are
+    // indexed from 1 to N, however the data locations are stored in the
+    // dataLocations array indexed from zero.
+    Token<OzoneBlockTokenIdentifier> token = keyLocationInfo.getToken();
+    BlockID blockID = keyLocationInfo.getBlockID();
+
+    Pipeline pipeline = keyLocationInfo.getPipeline();
+
+    List<DatanodeDetails> nodes = pipeline.getNodes();
+    List<DatanodeDetails> newNodes = new ArrayList<>();
+    ECReplicationConfig repConfig = (ECReplicationConfig)
+        pipeline.getReplicationConfig();
+    int totalNodes = repConfig.getRequiredNodes();
+    int parity = repConfig.getParity();
+
+    // Filtering the nodes that has the checksumBytes
+    for (int i = 0; i < nodes.size(); i++) {
+      if (i > 0 && i < totalNodes - parity) {
+        continue;
+      }
+      newNodes.add(nodes.get(i));
+    }
+
+    pipeline = Pipeline.newBuilder(pipeline)
+        .setReplicationConfig(StandaloneReplicationConfig
```

Comment:
Not familiar with EC implementation. Why is standalone type used instead of EC type?

Reply:
EC uses STANDALONE pipeline for reads. We create a STANDALONE pipeline that contains the location for the block index we want to read.